### PR TITLE
fixed #15781: Potential memory issue in StringUtil::format if argument size is greater than 4096

### LINKED ISSUE #15781: Potential memory issue in StringUtil::format if argument size is greater than 4096
https://github.com/cocos/cocos-engine/blob/develop/native/cocos/base/StringUtil.cpp#L89C13-L89C17

### PR DESCRIPTION
**Pull request by dumganhar**
Re: #15781

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: native/cocos/base/StringUtil.cpp [hunk 1]
```diff
@@ -86,12 +86,19 @@ int StringUtil::printf(char *buf, const char *last, const char *fmt, ...) {
 }
 
 ccstd::string StringUtil::format(const char *fmt, ...) {
-    char sz[4096];
     va_list args;
     va_start(args, fmt);
-    vprintf(sz, sz + sizeof(sz) - 1, fmt, args);
+    // Caculate the string length
+    int length = vsnprintf(nullptr, 0, fmt, args);
+    if (length <= 0) {
```
Comment (minggo) [inline native/cocos/base/StringUtil.cpp:93]: Does passing 0 will return the string length? From https://cplusplus.com/reference/cstdio/vsnprintf/, I don't find the information.

File: native/cocos/base/StringUtil.cpp [hunk 2]
```diff
@@ -86,12 +86,19 @@ int StringUtil::printf(char *buf, const char *last, const char *fmt, ...) {
...
+        return "";
```
Comment (minggo) [inline native/cocos/base/StringUtil.cpp:95]: Should invoke `va_end ` before return.

File: native/cocos/base/StringUtil.cpp (hunk 1 above)
Comment (dumganhar) [reply native/cocos/base/StringUtil.cpp:93]: Yes, it's for calculating the length of the final string. 
It works on macOS and windows. For other platforms, we need to run the `auto test`.

@cocos-robot run test cases

File: native/cocos/base/StringUtil.cpp (hunk 2 above)
Comment (dumganhar) [reply native/cocos/base/StringUtil.cpp:95]: yep, good catch.

Comment (dumganhar) [reply native/cocos/base/StringUtil.cpp:95]: FIXED

Comment (minggo) [reply native/cocos/base/StringUtil.cpp:95]: My be it is better to modify like
```c++
ccstd::string ret;
if (length > 0) {
   ...
}
...
return ret;
```

File: native/cocos/base/StringUtil.cpp (hunk 1 above)
Comment (minggo) [reply native/cocos/base/StringUtil.cpp:93]: It is better to add a unit test for this function.

Comment (minggo) [reply native/cocos/base/StringUtil.cpp:93]: Oh, just saw the unit test.

**Comment by dumganhar** [comment]
@cocos-robot run test cases

File: native/cocos/base/StringUtil.cpp (hunk 1 above)
Comment (dumganhar) [reply native/cocos/base/StringUtil.cpp:93]: I found a FreeBSD version of vsnprintf.c which is opensource by Apple , you could refer to 

https://github.com/apple-oss-distributions/Libc/blob/7861c72b1692b65f79c03f21a8a1a8e51e14c843/stdio/FreeBSD/vsnprintf.c#L73

Comment (dumganhar) [reply native/cocos/base/StringUtil.cpp:93]: It works on android too. Need to check whether it works on openharmony later.

**Comment by dumganhar** [comment]
@cocos-robot run test cases

File: native/cocos/base/StringUtil.cpp (hunk 1 above)
Comment (dumganhar) [reply native/cocos/base/StringUtil.cpp:93]: Works on OpenHarmony, too. Thanks @qiuguohua  for testing.

**Comment by dumganhar** [comment]
@cocos-robot run test cases

**Comment by sushanhong** [comment]
@cocos-robot run test cases

**Comment by dumganhar** [comment]
@cocos-robot run test cases

**Review by minggo — APPROVED** [review]

**Pull request merged**